### PR TITLE
7660 Avoid CI stability checks timing out

### PR DESCRIPTION
**Pull request by DanielRyanSmith**
addresses issue #7660

This is implemented based on [the proposed solution](https://docs.google.com/document/d/1dAlCSHUQldtgWDDTrGJR-ksm19FZZ3k8ppqc5-kSwIk/edit#) described by @stephenmcgruer.

Changes:
- Adds a new `--repeat-max-time` flag that will set how many minutes wptrunner should attempt to run the iterations of the test suite for the stability checks. Defaults to 100 minutes.
- When it is anticipated that an additional iteration run will eclipse this max time, the iterations of the test suite will terminate early.
- The test results will be evaluated based on the number of iterations run rather than the number attempted (10 by default). **As long as at least 2 or more iterations were completed.
- `wptrunner.run_tests` has had some refactoring to trim down the size of the function to be more readable.

### CODE REVIEW

File: css/CSS2/borders/border-001.xht
Comment (DanielRyanSmith) [inline css/CSS2/borders/border-001.xht:12]: Used to trigger the stability checks to actually run. Can be removed.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (stephenmcgruer) [inline tools/wptrunner/wptrunner/wptrunner.py:440]: Vague musing; I wonder if we should 'pad' the iteration time by e.g. +10%, to avoid the case where we estimate we will complete but actually narrowly avoid it.

**Review by stephenmcgruer — COMMENTED** [review]
This generally LGTM as an implementation of my old design. If ok, I would rather leave it to current WPT maintainers to do a 'real' review of the code, as they will be the ones maintaining it long term.

Side-note: there's a bunch of changes here that seem to be the result of some autoformatting happening in your editor. It would be great to revert those changes, as (a) they're inconsistent with WPT styling elsewhere I think (though we don't have a real style guide afaik...) but more importantly (b) they make it much harder to spot which deltas are new and which are just reformats.

File: tools/wptrunner/wptrunner/stability.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/stability.py]: I agree with this - definitely a clearer naming convention

File: tools/wptrunner/wptrunner/stability.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/stability.py]: I noticed that there is an inconsistency here, where it is always called with 10 runs, but the default is set to 5. Regardless, even if this is an inconsistency - this is not really the place to change it and should be separate. I've reverted it back.

File: tools/wptrunner/wptrunner/stability.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/stability.py]: Good idea - changed!

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: I do think this is a good idea and not terribly hard to implement - I've added the change.

File: tools/wptrunner/wptrunner/stability.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/stability.py]: Yes, I was hesitant about taking this approach. I know we definitely want to have access to the number of actual iterations that were run, but returning that information from wptrunner's `run_tests` implementation in a circumstance that it is only used when called by stability tests didn't feel right either. I am totally open to changing this implementation if anyone has suggestions or thinks a rework of `run_tests` is worth the change.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:388]: Please avoid using line continuation characters: it's preferred to use parens i.e.
```py
if (not repeat_until_unexpected and max_time
    and estimate >= start_time + max_time):
    # Condition body
```

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:391]: Although there's one existing example of the "a problem occured!" style of error mssage, I think I'd prefer to avoid that if possible. Also at risk of being needlessly pedantic, the formatting here which breaks things into three lines when they could be two seems needlessly wasteful of vertical space.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:394]: This could definitely be one or two lines, and could use a f-string (since we're touching the code anyway).

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:405]: I don't really know why this ended up getting reformatted, by my count the original line was 78 columns (I think we enforce 100 columns in pyflakes, but certainly more than this).

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:418]: Not sure why this got reformatted.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:426]: Padding here seems weird. Why not just pad when we compare to the remaining time?

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:180]: So a bunch of the reformatting here feels unnecessary, and this use of a continuation line in particular is not in keeping with the general project style.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:249]: No need to split this across multiple lines.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (jgraham) [inline tools/wptrunner/wptrunner/wptrunner.py:439]: I think we should make `run_tests` return `counts` directly. Then `start` can discard the extra information and we can avoid putting the output in a kwarg.

**Review by jgraham — CHANGES_REQUESTED** [review]
Some general comments:
* If you're refactoring code it's really helpful to have that in a seperate commit first and then the actual changes in a commits, such that the commits are reviewable one-at-a-time. Of course that doesn't always work out, but in this case I think it could have done.
* It looks like some kind of automated code formatting was applied. In general we've relied on pyflakes and not adopted automatic formatting (we couldn't enforce it anyway, since the project is used across multiple repos, and asking them to all use a specific version of a specific formatter is implausible). Also we're not aiming for 80 column lines; breaking before 100 is preferred.

In terms of the actual change; I think it's mostly good, but @stephenmcgruer is right (as usual) that using the kwargs as in in/out parameter feels like a bad idea. Even though we do a reasonable amount of in-place manipulation of the kwargs before passing them to `run_tests`, they aren't currently used for output. I think it will work out OK to update the function signature to return the extra data we need, and just drop it in `start` so we maintain compatibility with tools using that entry point.

Thanks for working on this!

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: I see, I'll approach it like this in the future!

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: I will remove this output and change the formatting 👍

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: Formatting fixed

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: This was probably my change when I reconfigured the code to use a counts dictionary. I'll make the formatting change.

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: I'll remove all instances of these line continuations - noted to not use these in the future!

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: switched to 1-liner 👍

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: Yes, looking at this possible solution, it's a relatively simple fix that can be easily implemented and will not affect other tests using wptrunner. I made a change to only return `counts["repeat"]` to only give back the number of iterations of the test suite that were run, rather than all of `counts`. Do you think this is a good idea? Or would it be functionally better to return all of `counts` in the chance that other uses might arise and need further access to the other counts?

File: tools/wptrunner/wptrunner/wptrunner.py
Comment (DanielRyanSmith) [reply tools/wptrunner/wptrunner/wptrunner.py]: Yes, I agree. I think it makes much more sense to store the actual runtime (as the variable states!) and pad the numbers only upon the comparison. Good change!

**Comment by DanielRyanSmith** [comment]
Thank you @stephenmcgruer and @jgraham for the reviews - I tried to remove some of the reformatting and I'll be sure to avoid any unnecessary auto-formatting changes and keep code refactoring in separate commits from now on.
Also, I would like to say that I appreciate how thorough the critiques of the code are, as it's helping me learn how to approach these issues moving forward!

**Review by jgraham — COMMENTED** [review]
This is looking pretty reasonable now, but I think at least returning the whole of `counts` (or even a new `Status` class like a static version of `counts`) as the second argument is preferable to just returning the number of iterations, since it gives us a way to return richer data in the future without making breaking changes to the API.

**Comment by DanielRyanSmith** [comment]
@jgraham, I tried to implement as you suggested. Normally I would create a dataclass for this, but since I we need to support Python 3.6 and dataclasses were added in 3.7, I just created a class with attributes to return. Maybe there is a better way to handle this?

https://github.com/web-platform-tests/wpt/blob/9d27581be56b4280801d3e87c8fe2ff6b31a3ac1/tools/wptrunner/wptrunner/wptrunner.py#L273-L281

What would @jgraham do in this scenario? (I imagine I will be asking this often, if only to myself sometimes 😄 )

**Comment by jgraham** [comment]
> What would @jgraham do in this scenario? 

I apologise in advance for any kind distress caused by thinking along those lines ;)

In this case I wonder if we could just have the `TestStatus` class and remove `counts` entirely, since we're not using the ability to add arbitary keys for anything other than convenience. So we'd change e.g. `counts["skipped"] += 1` to `status.skipped += 1`. I agree that using a dataclass will be nice once we can depend on 3.7+.

**Review by jgraham — APPROVED** [review]
Thanks! I think this looks good now. If the CI passes feel free to squash and merge with an appropriate commit message. If the decision task fails again we'll need to investigate what's going on; it seems unlikely to be this specific changeset.

**Pull request merged**
